compiler/src/dotty/tools/dotc/transform/TailRec.scala

```diff
@@ -233,15 +264,40 @@ class TailRec extends MiniPhase {
           if (ctx.tailPos) {
             c.debuglog("Rewriting tail recursive call:  " + tree.pos)
             rewrote = true
-            def receiver =
-              if (prefix eq EmptyTree) This(enclosingClass.asClass)
-              else noTailTransform(prefix)
-
-            val argumentsWithReceiver =
-              if (this.method.owner.isClass) receiver :: arguments
-              else arguments
 
-            tpd.cpy.Apply(tree)(ref(label), argumentsWithReceiver)
+            val assignParamPairs = for {
+              (param, arg) <- paramSyms.zip(arguments)
+              if (arg match {
+                case arg: Ident => arg.symbol != param
+                case _ => true
+              })
+            } yield {
+              (getVarForRewrittenParam(param), arg)
+            }
+
+            val assignThisAndParamPairs = {
+              if (prefix eq EmptyTree) assignParamPairs
+              else {
+                // TODO Opt: also avoid assigning `this` if the prefix is `this.`
+                (getVarForRewrittenThis(), noTailTransform(prefix)) :: assignParamPairs
+              }
+            }
+
+            val assignments = assignThisAndParamPairs match {
+              case (lhs, rhs) :: Nil =>
+                Assign(ref(lhs), rhs) :: Nil
+              case _ :: _ =>
+                val (tempValDefs, assigns) = (for ((lhs, rhs) <- assignThisAndParamPairs) yield {
+                  val temp = c.newSymbol(method, TailTempName.fresh(lhs.name.toTermName), Flags.Synthetic, lhs.info)
+                  (ValDef(temp, rhs), Assign(ref(lhs), ref(temp)).withPos(tree.pos))
+                }).unzip
+                tempValDefs ::: assigns
+              case nil =>
+                Nil
+            }
+
+            val tpt = TypeTree(method.info.resultType)
+            Block(assignments, Typed(Return(Literal(Constant(())).withPos(tree.pos), ref(getContinueLabel())), tpt))
```

Comment:
Why is the `Return` tree wrapped in a `TypedTree` node?

Reply:
To perfectly preserve the type of the node (otherwise it's `Nothing`), which in turn is very important for the `TypeAssigner` not to infer different types as lubs of branches/cases when one of the branches has been replaced by a `return`. Otherwise Ycheck failures arise.